include/deal.II/base/memory_space_data.h

```diff
@@ -25,12 +25,124 @@
 #include <functional>
 #include <memory>
 
+#ifdef DEAL_II_USE_KOKKOS_BACKEND
+#  include <Kokkos_Core.hpp>
+#endif
+
 DEAL_II_NAMESPACE_OPEN
 
 /**
  */
 namespace MemorySpace
 {
+#ifdef DEAL_II_USE_KOKKOS_BACKEND
+  /**
+   * Data structure
+   */
+  template <typename Number, typename MemorySpace>
+  struct MemorySpaceData
+  {
+    MemorySpaceData()
+      : values("memoryspace data", 0)
+    {}
+
+    bool
+    has_data()
+    {
+      return values.extent(0) > 0;
+    }
+
+    Number *
+    data()
+    {
+      return values.data();
+    }
+
+    const Number *
+    data() const
+    {
+      return values.data();
+    }
+
+    Number &
+    operator()(unsigned int i)
+    {
+      return values(i);
+    }
+
+    const Number &
+    operator()(unsigned int i) const
+    {
+      return values(i);
+    }
+
+    /**
+     * Copy the active data (values for Host and values_dev for CUDA) to @p begin.
+     * If the data is on the device it is moved to the host.
+     */
+    void
+    copy_to(Number *begin, std::size_t n_elements)
+    {
+      Assert(n_elements <= values.extent(0), ExcMessage("aa"));
+      Kokkos::parallel_for(
+        "MemorySpaceData::copy_to",
+        Kokkos::RangePolicy<typename MemorySpace::execution_space>(0,
+                                                                   n_elements),
+        KOKKOS_LAMBDA(const int i) { *(begin + i) = values(i); });
+    }
+
+    /**
+     * Copy the data in @p begin to the active data of the structure (values for
+     * Host and values_dev for CUDA). The pointer @p begin must be on the host.
+     */
+    void
+    copy_from(Number *begin, std::size_t n_elements)
+    {
+      Assert(n_elements <= values.extent(0), ExcMessage("bb"));
+      Kokkos::parallel_for(
+        "MemorySpaceData::copy_from",
+        Kokkos::RangePolicy<typename MemorySpace::execution_space>(0,
+                                                                   n_elements),
+        KOKKOS_LAMBDA(const int i) { values(i) = *(begin + i); });
+    }
+
+    /**
+     * Kokkos View to the data
+     */
+    Kokkos::View<Number *, MemorySpace> values;
+
+    /**
+     * Pointers to the data of the processes sharing the same memory.
+     */
+    std::vector<ArrayView<const Number>> values_sm;
```

Comment:
I have to admit that I don't understand what the comment refers to. I suspect that's because I don't understand what the semantics of this object are supposed to be.